### DISCUSSION TITLE
Welcome to AndroidAutoIdrive Discussions!

### DISCUSSION THREAD

**Discussion by hufman** [BimmerGestalt/AAIdrive#152] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

Feel free to start a new discussion about any topic!

**Comment by mwekell** [comment]
This app is very cool, I have tried it out and it works on my BMW X6 2016 f16. My understanding is that audio quality would be higher via USB, I believe Spotify ends up going over Bluetooth. Is that correct ? Can I play Spotify over the USB cable? Thank you for your help!

**Reply by hufman** [reply]
You are correct, unfortunately! As far as I know, the car implements audio over USB that is only compatible with the iPhone protocol and the legacy AOAv2 protocol that was removed in Android Oreo. Any newer Android versions can only send sound through Bluetooth.

**Comment by gabrik1** [comment]
Congratulations for this app, and to give a oportunity to Android lovers in BMW/MINI environmets.
I see taht recently, the MINI connected app has update to MINI app, and it has a new UI firmware. with these changes, do your app still working?

**Reply by hufman** [reply]
Sadly, the new MINI app is related to the new My BMW app, which seems to drop support for the protocol that my app uses: #149

**Reply by gabrik1** [reply]
thanks for your support hufman. your app was a very good option tu have android auto in this vehicles :( you are working on it? or this is other war.

**Reply by hufman** [reply]
I don't think there's currently any way with the new apps to achieve the same functionality, of allowing phone apps to display information remotely in the car. The apps themselves contain most of the necessary code for it, and my app should just work if they do enable it. It's possible they've only disabled it temporarily during this early development while they are adding other features and that it may come back!

**Reply by gabrik1** [reply]
thanks hufman, and good job :)

**Comment by mwekell** [comment]
Thank-you very much for writing back, I greatly appreciate. Excellent app!!
I am actually curious to run it on my android head unit BMW interface
https://www.amazon.com/dp/B087286YBW/ref=cm_sw_r_cp_apa_fabc_p3x3Fb6VK0ETQ.
Usb from the headunit to the USB Idrive, my goal would be if I selected a
Google maps destination that it could be automatically sent to the BMW
navigation to make use of my heads up display. Any ideas on this ?

Thank you again!!

**Reply by hufman** [reply]
That's a clever integration idea! I don't know of a way to discover the Google Maps navigation destination, but if that were figured out it would seem possible to automatically forward this destination to the car's navigation. The app knows how to trigger the car's navigation, so half of the work is done! 😃

**Comment by Tyfon1** [comment]
Hi, first of, great app! Tried it for the first time today, had some struggling with the connection, but now it seems to be doing great.

Just one question, it says that also Google Maps is implemented, but I cant seem to find out how?

Thx!

**Reply by hufman** [reply]
You are correct! There is a code module that can draw a Google Mapview, search using Google Places, and draw a navigation line using Google Directions. It is not strictly the same as Google Maps: It does not have your search or place history, automatic rerouting is not implemented, and in fact it's strictly against the EULA to replicate the Google Navigation experience at all. This module is not enabled by default because of Google Map API restrictions, and so you must follow the instructions at the bottom of the README for how to set up the Google API token and compile the app with this module enabled.

I'm also not satisfied with its performance, averaging ~1fps over Bluetooth and a slightly better 10fps over USB while also starting the phone on fire with CPU usage. I can't recommend going through the trouble of custom compiling the app for fear you'll be disappointed by the result :)

A replacement mapping service is on the roadmap for version 1.4, based on Mapbox which has no such restrictions. I fear the performance will still be bad, so please don't hold your breath :)

**Comment by jmone1** [comment]
Thanks for your work on this!!  I can't believe that in 2020/21 the Mini still does not have Android Auto :(  Anyway, I'm trying to compile with my Google API Key and have 
- installed Android Studio (4.1.1) for Windows and configured it to only have the Android 9.0 (API Level 28) SDK Platform Installed
- downloaded the project from github into a folder
- edited 'gradle.properties' --> AndroidAutoIdrive_GmapsApiKey=MyKey
- added all the linked APK's to the external folder
-
but I get the following warnings and errors when running "./gradlew assembleNomapNonalyticsDebug"

```
> Configure project :app
Warning: The 'kotlin-android-extensions' Gradle plugin is deprecated. Please use this migration guide (https://goo.gle/kotlin-android-extensions-deprecation) to start working with View Binding (https://developer.android.com/topic/libraries/view-binding) and the 'kotlin-parcelize' plugin.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> android.defaultConfig.versionCode is set to 0, but it should be a positive integer.
  See https://developer.android.com/studio/publish/versioning#appversioning for more information.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```
Thanks in advance for any tips (first time using Android Studio)
Nathan

**Reply by hufman** [reply]
Good evening! This error is due to the build script not being able to find any git information in the folder, and so it can't figure out what version number to set in the APK.
Android Studio's terminal pane has git available on the path for convenient cloning even on Windows (I'm fairly certain, at least), or you can use `File > New > Project From Version Control` to have Android Studio clone the project for you.
I should fix the build script to fail more gracefully in this case! While I work on that, try out the git checkout method and see if that helps.

**Reply by jmone1** [reply]
Thanks - that worked and I could run up Google Maps / Navigation on the Mini!  That said, I agree with your comments that it is a lesser experience than the real Goggle Maps / Navigation.  I also noticed that I lost Spotify Control with my own complied version (I did not try to add a custom Spotify API Key).  I think I'll go back to your pre-compiled version and just run Google Maps / Navigation from my Android Phone when needed (sitting in the Cup Holder).  

Thanks again for your work, and I've sent some coffees your way.
Nathan

FYI - When asking my Mini Dealer when we were getting Android Auto (as BMW now has it), I was told it would be at least a year away.

**Comment by sawinco** [comment]
Great work so far, just a couple of quick questions. If I use Spotify via the mini connected app, can I use amazon music in your media a player. With amazon music should I get cover art as I don't at present but do with Spotify, Deezer via your app. At present I use tasker to launch Mini connected app when connected to car Bluetooth which in turn launches Spotify. Will adapt to launch your app which could in turn launch amazon music.
Thanks Simon

**Reply by hufman** [reply]
Yes, this app uses the same protocol that Spotify and Mini Connected use, so if they work already, this app will work too. Most music apps generally allow this app to load the cover art too, but it is up to the app to provide it.
Note, however, that Amazon Music doesn't allow this app to launch it, so you'll need to set up Tasker to also start Amazon Music.

**Reply by sawinco** [reply]
Thanks for the reply so do you think Amazon music should supply cover art back to your app. Is the AVRCP version relevant for your app? Plus in the musicplayer with cover art e. g Spotify is there a way to make the cover art bigger.
Many thanks Simon

**Reply by hufman** [reply]
I do think it _should_, but I just tested and Amazon Music did not provide cover art for me. AVRCP version is unimportant, the coverart is shown in the car with the BMW Apps protocol and the coverart is fetched from the music app directly via the Android MediaController API.
The Audioplayer interface is locked to 200x200px coverart, or 320x320px if the car is running in wide-screen mode, and unfortunately can not be changed.

**Reply by sawinco** [reply]
Hi just a bit more feedback. Using the latest nightly build I am using alexa to play music, so neither Spotify or amazon music installed. This does seem to eventually show cover art if I scroll back a track the art work shows up.
Still trying things out but the reason I want to use amazon music in my mini is because I have amazon prime so can hopefully drop Spotify premium. Looks promising so far thanks for all your work

**Comment by fookyuanloh** [comment]
Would you make this app avail over google appstore?

I'm having issues to install this - mobileiron for work emails/profile etc is preventing me to install any side load apks
Something to do with permission config by the admin when the mobileiron is setup, to prevent "unverified" apks

Any plans to make it avail over appstore?

**Reply by hufman** [reply]
Yes! With all the work in this upcoming 1.3.0 release (which was admittedly too much and took far too long) to make the app more self-explanatory, this unlocks the ability to post the app to the app store. I filed #212 to track it just yesterday in fact.

**Reply by fookyuanloh** [reply]
Cool
Will wait for the next release

**Comment by jpmreece** [comment]
This app has been a lifesaver in terms of regaining functionality! Thank you.

As a dev, I like to build and integrate myself but have had problems with the Spotify integration. I have created the App Id and redirected the URI but I get an invalid account error message. Are you able to help?

Thanks
Jon

**Comment by Petran911** [comment]
Hello, first of all congratulations for this app. Is it possible that you help those of us without so much knowledge related to Android Coding about the Google map api?  Followed all the instructions, compiled apk but I am getting the google protect error in my mobile.

**Reply by waldemar-szostak** [reply]
Have u tried just installing the compiled apk?

**Reply by hufman** [reply]
I can't offer much help about Google Play Protect, the Android Studio should automatically configure the development certificate and it should install properly.

**Comment by dlmarien** [comment]
@hufman Hello again! I have a question, but I'm reasonably certain I know the answer. My wife's 2018 X1 subscription to BMW Connected ran out mid-April (2021). The subscription is free for 3 years, then you must re-subscribe annually. The annual cost is ~$200US, but before I do this, I wanted to verify that w/o the subscription, the "Apps" check box will not appear and therefore apps such as Spotify will not appear either. Am I correct in saying BMW Connected sub is required to generate the Apps check box? I hope I've explained this correctly. Thank you for your help!

**Reply by fabssor** [reply]
So I don't think an active subscription of a connected package is required. The subscription is only needed for the BMW online services like locating the car, sending routes to the car, etc. (depending on the level of subscription additional feature are added).

This is at least the case for Germany :)

**Comment by dlmarien** [comment]
Thanks for the reply. In which case, this begs the question what am I not doing correctly in order to display the App option / Spotify? I'm running both old & new BMW app, latest build for AAI and latest version of Spotify. Android 11 on Pixel 3XL....I have deleted the mobile from iDrive as well as the BT connection fr the mobile...clearly I'm missing something assuming the Sub is not a requirement

**Reply by fabssor** [reply]
Did it work before the subscription ran out? Be sure your car has the following optional accessories SA 6NR or SA 6AK. Those must be optional ordered when you have bought your car or somebody must code them afterwards into your car.
Also be sure you phone is connected as primary phone (if I remember correctly apps are only available as primary phone but maybe I am wrong here).
Also follow the guide from hufman maybe this helps out https://hufman.github.io/AndroidAutoIdrive/faq.html

**Comment by dlmarien** [comment]
Hello. Yes, the App option previously was available for both mobile devices (SG10+ and my Pixel 3XL). I've been out to her vehicle just now, deleted the mobile from her vehicle (ID6) and still no App option appears. I've tried this with her SG10+ as well. I have so tried this with both BMW apps installed on the mobile phones and still no go...it's not critical, was hoping there'd be a reasonably good answer why it may have disappeared but more importantly why the App option isn't returning. Oh yes, I double checked hufman's FAQ's as well

**Reply by fabssor** [reply]
That's very strang. Maybe this will helop:
https://www.bimmerfest.com/threads/lost-online-functions-from-connecteddrive-after-coding.888708/#post-9452582

**Reply by dlmarien** [reply]
Thanks, I hadn't thought of that. Selected the option to update services then logged out and back in again. No luck unfortunately but thank you for trying! I guess perhaps in Canada/US a subscription may be required after all. I'll try to contact the Connected team to see if they can assist...so bizarre as it worked perfectly a month ago or thereabouts

**Comment by dlmarien** [comment]
Good morning! This is why I had the impression a current Connected Drive subscription is required for the "App" option to appear in iDrive --> Any MY2014 or newer BMW or Mini equipped with NBT or NBT Evo and the "BMW Apps (6NR)" feature, **an active BMW ConnectedDrive subscription**, or the "Mini Connected (SA6NM)" option should be compatible. Furthermore, it could explain why it was working 1+ month(s) ago, but not now, since the subscription on her vehicle expired 3 years after purchase, which was ~19th April. Hoping someone can confirm this is the case, at least for North America. @hufman ? What do you think?

**Reply by bogdan-calapod** [reply]
Isn't the ConnectedDrive subscription required for *online* services, like turning on/off the headlights and stuff ?

**Reply by dlmarien** [reply]
You are correct but it seems coincidental that her sub expired a month ago or so, and that's when the App option in iDrive disappeared...I just can't explain why the App option is no longer available. Must be the lack of sub, no?

**Reply by bogdan-calapod** [reply]
Ah I just now saw that there are multiple threads open on this topic - please let's keep a single topic for each issue 😅 Ideally we should open a new discussion (instead of here where it is the "intro/hello" discussion topic).

The App option is local only, I really doubt that it should be affected by the subscription state.

**Reply by dlmarien** [reply]
Apologies on the thread. I'll track down the existing thread / discussion. Thanks!

**Reply by hufman** [reply]
I'm sorry to hear that! I had added that note because the Connected Drive subscription seemed to help the ~5 people who reached out over email over the past year. Likewise, as you noticed, the subscription expired and the Apps option disappeared around the same time. I think there's a confounding factor of a delay between the subscription status and the car's Apps status, with one (Germany) user needed to wait overnight before the car noticed the new subscription, and someone else pointing out an Options menu to force it to reload subscription status. Another recent user was the one to discover and tell me that BMW Connected 6.5 stopped working with the car, and he indicated that downgrading to Connected 6.4 brought the Apps icon back to the car, but I have no idea why.

It's frustratingly difficult to find exact recommendations, BMW has stopped mentioning this BMW Apps feature (probably because IDrive7 doesn't support it) and so their information sites no longer say what is required for the Apps functionality.

I unfortunately can't confirm from experience directly, my Mini doesn't have a subscription service and I believe relies solely on the car's coded SA6NM option.

**Reply by dlmarien** [reply]
Hi @hufman thank you for the reply. I have not tried downgrading to cinn cted version 6.4.xxx because I cannot seem to find a mirror site for this app for US/Canada would you be able to point me in the right direction? Google search brings a limited number of mirror sites but none seem to be for Canada?

**Reply by dlmarien** [reply]
Found a site. Will retry in the vehicle now

**Reply by hufman** [reply]
:+1: There are some more links [in this guide](https://github.com/hufman/AndroidAutoIdrive/discussions/255). I presume Canada has the same version as the US, because the package code adds "na" which I presume means North America.

**Reply by dlmarien** [reply]
Thanks! I did in fact find one. However, I tried it and still no App option...oh well, works in my vehicle just can't explain why it was fine in hers before and now not...other than the expired sub...thanks everyone for trying

**Reply by dlmarien** [reply]
I can confirm (for Canada) that a subscription to Connected Drive services is required for the "App" function to work. I updated my wife's subscription yesterday and voila...the Smartphone App functionality has returned (along with the other expected options such as Weather, RTTI etc)

**Reply by hufman** [reply]
Thank you for following up! I'm glad it's working again!

**Comment by DoctorHub02** [comment]
Hello! First of all, thank you for the great App and the new integration for MyBMW! Just bought a 1er 2017 and was disappointed that Spotify ditched the support on Android. Your app was life saviour. But I have a question: I can't browse my library in Spotify because it requires explicit authorization or something on those lines. Where is the option? Thanks!

**Reply by hufman** [reply]
If you open up the Music tab in the phone app, it should show an empty checkbox with an button to Grant the `Ability to browse your Spotify library`. The Spotify SDK will then go through the authorization flow (or return any errors) and then browsing should work in the car. You can click on the Spotify app in that list to open a test PhoneUI music player to try out the browse functionality without needing to go out to the car.

**Reply by DoctorHub02** [reply]
Thanks for the answer. When I open the Music Tab, I can't check the box cause it gives error says something about auth-flow and explicit authorization. How can I solve the problem? I'm on 1.31 release

**Reply by hufman** [reply]
The error comes from the Spotify SDK, which is pretty picky about the installed Spotify version. It refuses to support any modded Spotify version, for example. I think that generally returns a different error though, the Explicit Authorization seems like it's complaining that it can't open the popup window to grant access for some reason. I am unable to give clear answers, I'm sorry: is Spotify running, I assume it's signed in properly, is it running in Online mode instead of Offline mode...

**Reply by DoctorHub02** [reply]
Thanks for the clarification. Now for some unknown reason (giving all permission to Spotify, Android Idrive and of course MyBMW) I'm able to check the box and make it fully compatible. Love how technology works ahahaha

**Reply by hufman** [reply]
Great, that's good to hear!

**Comment by harnas81** [comment]
First of all, thanks for great APP! Could you tell, is somewhere to download version with added gmap? I think im so fresh to edit or im doing something wrong. I cant compile apk.

**Reply by hufman** [reply]
I'm glad you like it! I can't provide a GMaps-enabled build, because it would would violate the Google Maps Terms of Service to show the map in the car. Later this year, there is planned to be a version that adds a map interface from Mapbox.

**Comment by minmak11** [comment]
Hi,
The app is amazing for us who don't have iphones. I worked most of the functions, but navigation doesn't work.
Any suggestions/solution?
I'm running the latest version of the app on a Google Pixel 6 pro. Mini cooper s.

**Reply by hufman** [reply]
Thank you for the compliments! Can you tell me more about what isn't working? Can you share the exact build version you are using? What are you expecting it to do, and what it is doing differently? If you mean Turn By Turn Guidance, that is not implemented (#539).

**Reply by minmak11** [reply]
Yes, i meant guidance. Any possibility of implementing this feature?

**Reply by hufman** [reply]
Everything is possible, of course :) However, I don't have any good ideas on how to implement this feature, so I won't be able to build it anytime soon.

**Comment by catarinalopes27** [comment]
Hello, loved to hear about the app! Does it work with the newest Mini's ID6? I can't get it to connect to the car for some reason. Thank you!

**Reply by hufman** [reply]
As far as I know, Mini ID6 is compatible. Does your car show the Apps option in the Mobile Phone menu? Does the Mini app show the Apps connection?

**Reply by catarinalopes27** [reply]
Yeah, my car shows the Apps option. How do I check the Mini App for the Apps connection?

**Reply by hufman** [reply]
Within the Mini app, there should be a button at the bottom named Device Connection, which brings up this screen:
![photo1664392810](https://user-images.githubusercontent.com/1592375/192870316-8b4470b9-ad65-44c7-94b4-2bf9aaaa9a6f.jpeg)

**Reply by catarinalopes27** [reply]
That might be the issue. I have a button that should set up alexa, but it's not working. It doesn't mention anything regarding Device connection though. Thanks for the help

**Reply by hufman** [reply]
Very interesting! What version of the Mini app do you have? I have version 2.9.1 and it shows the option for me.

**Reply by catarinalopes27** [reply]
It just updated to 2.9.2, but mine actually has all these options in the bottom
![Screenshot_20220928_213049](https://user-images.githubusercontent.com/94445494/192882907-7b4cb626-1bc2-4f9d-814f-a1e0b8992b80.jpg)

**Reply by catarinalopes27** [reply]
Could it be because mine is Portuguese?

**Reply by catarinalopes27** [reply]
Ok, I tried it again today and it worked! It might be because I installed the Mini Connected apk yesterday. Even though the app doesn't work, somehow it made the connection.

**Comment by Mesiyas** [comment]
Hi Hufman, thank you so much for this amazing app. I love it 😍 
I have a question about: I wanted to ask you if it would be possible to add Replaio Radio App too?
I mean just like My Tuner, for example. With searchable favourites like in my tuner. That would be awesome. Thanks again for your amazing work

**Reply by hufman** [reply]
No, unfortunately Replaio enforces a list of which apps are allowed to connect and control it. Please reach out to the Replaio developer to ask them to allow any apps to connect to its MediaBrowserService.
